InformationScripting/src/parsing/QueryBuilder.cpp

```diff
@@ -0,0 +1,209 @@
+	for (QString part : parts)
+	{
+		auto type = typeOf(part);
+		if (Type::Operator == type)
+			result.push_back(parseOperator(part, true));
+		else if (Type::Query == type)
+			result.push_back(parseQuery(part));
+		else
+			Q_ASSERT(false); // TODO error for user
+	}
+	return result;
+}
+
+Query* QueryBuilder::parseOperator(const QString& text, bool connectInput)
+{
+	// TODO it might be better to be able to register operators.
+	// But that need some additional information on how they are used, thus we currently hardcode the operators.
+	auto splitted = split(text, {'|', '-', 'U'});
+	auto parts = splitted.first;
+	auto operators = splitted.second;
+	Q_ASSERT(parts.size()); // TODO error for user
+	CompositeQuery* composite = new CompositeQuery();
+	auto previousQueries = parseOperatorPart(parts[0]);
+	if (connectInput)
+	{
+		for (int i = 0; i < previousQueries.size(); ++i)
+			composite->connectInput(i, previousQueries[i]);
+	}
+
+	for (int i = 1; i < parts.size(); ++i)
+	{
+		auto currentQueries = parseOperatorPart(parts[i]);
+		// If left and right are both lists we don't know how to map things:
+		Q_ASSERT(!(previousQueries.size() > 1 && currentQueries.size() > 1)); // TODO error for user
+		if (previousQueries.size() == 1 && currentQueries.size() == 1)
+		{
+			if (operators[i-1] == '|')
+				composite->connectQuery(previousQueries[0], currentQueries[0]);
+			else if (operators[i-1] == '-')
+			{
+				// TODO maybe minus should look like a union op, i.e. multiple input single output?
+				auto minus = new SubstractNodesOperator();
+				composite->connectQuery(previousQueries[0], minus);
+				composite->connectQuery(currentQueries[0], 0, minus, 1);
+				currentQueries = {minus};
+			}
+		}
+		else if (previousQueries.size() > currentQueries.size())
+		{
+			auto unionOp = new UnionOperator();
+			for (int i = 0; i < previousQueries.size(); ++i)
+				composite->connectQuery(previousQueries[i], 0, unionOp, i);
+			composite->connectQuery(unionOp, currentQueries[0]);
+		}
+		else
+		{
+			for (auto currentQuery : currentQueries)
+				composite->connectQuery(previousQueries[0], currentQuery);
+		}
+		previousQueries = currentQueries;
+	}
+	// It might be that we have an operator at the end
+	if (parts.size() > operators.size() || operators.last() == '|')
+	{
+		for (int i = 0; i < previousQueries.size(); ++i)
+			composite->connectToOutput(previousQueries[i], i);
+	}
+	else if (operators.last() == 'U')
+	{
+		auto unionOp = new UnionOperator();
+		for (int i = 0; i < previousQueries.size(); ++i)
+			composite->connectQuery(previousQueries[i], 0, unionOp, i);
+		composite->connectToOutput(unionOp);
+	}
+	else // No other case allowed
+		Q_ASSERT(false); // TODO error for user
+	return composite;
+}
+
+QList<Query*> QueryBuilder::parseOperatorPart(const QString& text)
+{
+	auto type = typeOf(text);
+	if (Type::Query == type) return {parseQuery(text)};
+	else if (Type::List == type) return parseList(text);
+	Q_ASSERT(false);
+}
```

Comment:
Hm have you looked at the grammar it should be able to parse in the header file? And after looking at it are you still sure we can solve it with regular expressions? If so I will be happy to redo it with regexes.